File: fake-store-quarkus/src/main/java/org/arielzarate/interfaces/rest/ProductController.java
package org.arielzarate.interfaces.rest;

import lombok.AllArgsConstructor;
import org.arielzarate.domain.model.Product;
import org.arielzarate.domain.ports.in.ProductService;
import org.arielzarate.interfaces.rest.dto.ProductRequest;
import org.arielzarate.interfaces.rest.dto.ProductResponse;
import org.arielzarate.interfaces.rest.mapper.ProductMapper;
import jakarta.inject.Inject;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;

import java.util.List;
import java.util.stream.Collectors;

@Path("/products")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@AllArgsConstructor
public class ProductController {


    private final ProductService productService;

    @GET
    public List<ProductResponse> getAllProducts() {
        List<Product> products = productService.getAllProducts();
        return products.stream()
                .map(ProductMapper::toResponse)
                .collect(Collectors.toList());
    }

    @GET
    @Path("/{id}")
    public Response getProductById(@PathParam("id") Integer id) {
        Product product = productService.getProductById(id);
        if (product == null) {
            return Response.status(Response.Status.NOT_FOUND).build();
        }
        return Response.ok(ProductMapper.toResponse(product)).build();
    }

    @POST
    public Response createProduct(ProductRequest request) {
        Product product = ProductMapper.toDomain(request);
        Product created = productService.createProduct(product);
        return Response.status(Response.Status.CREATED)
                .entity(ProductMapper.toResponse(created)).build();
    }
}

File: api-fake-hexagonal/src/main/kotlin/com/arielZarate/hexagonal_example/interfaces/ProductController.kt
package com.arielZarate.hexagonal_example.interfaces

import com.arielZarate.hexagonal_example.domain.model.Product;
import com.arielZarate.hexagonal_example.domain.ports.`in`.GetProductService
import com.arielZarate.hexagonal_example.interfaces.mappers.ProductMapper
import com.arielZarate.hexagonal_example.interfaces.model.ProductRequest
import com.arielZarate.hexagonal_example.interfaces.model.ProductResponse
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*


@RestController
@RequestMapping("/products")
class ProductController(
    private val getProductService: GetProductService,
    private val productMapper: ProductMapper
) {
    @GetMapping
    fun getAllProducts(): ResponseEntity<List<ProductResponse>> {
        val products: List<Product> = getProductService.getAllProducts()
        return ResponseEntity.ok(products.map { productMapper.mapToResponse(it) })
    }

    @GetMapping("/{id}")
    fun getProductById(@PathVariable id: Int): ResponseEntity<ProductResponse> {
        val product: Product? = getProductService.getProductById(id);
        return ResponseEntity.status(HttpStatus.OK).body(product?.let { productMapper.mapToResponse(it) })
    }

    @PostMapping
    fun createProduct(@RequestBody productRequest: ProductRequest): ResponseEntity<ProductResponse> {
        val product = productMapper.mapToDomain(productRequest)
        val createdProduct: Product = getProductService.createProduct(product)
        return ResponseEntity.status(HttpStatus.CREATED).body(productMapper.mapToResponse(createdProduct))
    }
}